libr/parse/p/parse_riscv_pseudo.c

```diff
@@ -0,0 +1,229 @@
+		{ 0, "sw", "[#] = #", { 2, 1 } },
+		{ 0, "sb", "[#] = #", { 2, 1 } },
+		{ 0, "sh", "[#] = #", { 2, 1 } },
+		{ 0, "sub", "# = # - #", { 1, 2, 3 } },
+		{ 0, NULL }
+	};
+	if (!newstr) {
+		return false;
+	}
+
+	for (i = 0; ops[i].op; i++) {
+		if (ops[i].narg) {
+			if (argc - 1 != ops[i].narg) {
+				continue;
+			}
+		}
+		if (!strcmp (ops[i].op, argv[0])) {
+			if (newstr) {
+				d = 0;
+				j = 0;
+				ch = ops[i].str[j];
+				for (j = 0, k = 0; ch != '\0'; j++, k++) {
+					ch = ops[i].str[j];
+					if (ch == '#') {
+						if (d >= MAXPSEUDOOPS) {
+							// XXX Shouldn't ever happen...
+							continue;
+						}
+						int idx = ops[i].args[d];
+						d++;
+						if (idx <= 0) {
+							// XXX Shouldn't ever happen...
+							continue;
+						}
+						const char *w = argv[idx];
+						if (w) {
+							strcpy (newstr + k, w);
+							k += strlen (w) - 1;
+						}
+					} else {
+						newstr[k] = ch;
+					}
+				}
+				newstr[k] = '\0';
+			}
+			r_str_replace_char (newstr, '{', '(');
+			r_str_replace_char (newstr, '}', ')');
+			return true;
+		}
+	}
+
+	/* TODO: this is slow */
+	newstr[0] = '\0';
+	for (i = 0; i < argc; i++) {
+		strcat (newstr, argv[i]);
+		strcat (newstr, (!i || i == argc - 1)? " " : ",");
+	}
+
+	r_str_replace_char (newstr, '{', '(');
+	r_str_replace_char (newstr, '}', ')');
+	return false;
+}
+
+static int parse(RParse *p, const char *data, char *str) {
+	char w0[256], w1[256], w2[256], w3[256];
+	int i, len = strlen (data), n;
+	char *buf, *ptr, *optr, *num;
+
+	if (len >= sizeof (w0)) {
+		return false;
+	}
+	// malloc can be slow here :?
+	if (!(buf = malloc (len + 1))) {
+		return false;
+	}
+	memcpy (buf, data, len + 1);
+	if (*buf) {
+		*w0 = *w1 = *w2 = *w3 = '\0';
+		ptr = strchr (buf, ' ');
+		if (!ptr) {
+			ptr = strchr (buf, '\t');
+		}
+		if (ptr) {
+			*ptr = '\0';
+			for (++ptr; *ptr == ' '; ptr++) {
+				;
+			}
+			strncpy (w0, buf, sizeof (w0) - 1);
+			strncpy (w1, ptr, sizeof (w1) - 1);
+
+			optr = ptr;
+			if (*ptr == '(') {
+				ptr = strchr (ptr+1, ')');
+			}
+			if (ptr && *ptr == '[') {
+				ptr = strchr (ptr+1, ']');
+			}
+			if (ptr && *ptr == '{') {
+				ptr = strchr (ptr+1, '}');
+			}
+			if (!ptr) {
+				eprintf ("Unbalanced bracket\n");
+				free(buf);
+				return false;
+			}
+			ptr = strchr (ptr, ',');
+			if (ptr) {
+				*ptr = '\0';
+				for (++ptr; *ptr == ' '; ptr++) {
+					;
+				}
+				strncpy (w1, optr, sizeof (w1) - 1);
+				strncpy (w2, ptr, sizeof (w2) - 1);
+				optr = ptr;
+				ptr = strchr (ptr, ',');
+				if (ptr) {
+					*ptr = '\0';
+					for (++ptr; *ptr == ' '; ptr++) {
+						;
+					}
+					strncpy (w2, optr, sizeof (w2) - 1);
+					strncpy (w3, ptr, sizeof (w3) - 1);
+				}
+			}
+			ptr = strchr (buf, '(');
+			if (ptr) {
+				*ptr = 0;
+				num = (char*)r_str_lchr (buf, ' ');
+				if (!num) {
+					num = (char *)r_str_lchr (buf, ',');
+				}
+				if (num) {
+					n = atoi (num+1);
+					*ptr = '[';
+					memmove (num+1, ptr, strlen (ptr)+1);
```

Comment:
spaces around + and suhc..but you can use r_str_cpy, which is doing memmove under the hood, but you dont have to use the strlen by hand